src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs

```diff
@@ -474,20 +471,68 @@ internal static string GetMSBuildExecutablePathForNonNETRuntimes(HandshakeOption
         /// - RuntimeHostPath: The path to the dotnet executable that will host the .NET runtime
         /// - MSBuildAssemblyPath: The full path to MSBuild.dll that will be loaded by the dotnet host.
         /// </returns>
-        internal static (string RuntimeHostPath, string MSBuildAssemblyPath) GetMSBuildLocationForNETRuntime(HandshakeOptions hostContext)
+        internal static (string RuntimeHostPath, string MSBuildAssemblyPath) GetMSBuildLocationForNETRuntime(HandshakeOptions hostContext, Dictionary<string, string> taskHostParameters)
         {
             ErrorUtilities.VerifyThrowInternalErrorUnreachable(Handshake.IsHandshakeOptionEnabled(hostContext, HandshakeOptions.TaskHost));
 
-            s_baseTaskHostPathNet ??= BuildEnvironmentHelper.Instance.MSBuildToolsDirectoryNET;
+            taskHostParameters.TryGetValue(Constants.DotnetHostPath, out string runtimeHostPath);
+            var msbuildAssemblyPath = GetMSBuildAssemblyPath(taskHostParameters);
 
-            string msbuildAssemblyPath = Path.Combine(BuildEnvironmentHelper.Instance.MSBuildAssemblyDirectory, Constants.MSBuildAssemblyName);
-            string runtimeHostPath = Path.Combine(s_baseTaskHostPathNet, Constants.DotnetProcessName);
-
-            // Current process is .NET Framework, but handshake requests .NET
-            // Launch dotnet host and path to MSBuild.dll
             return (runtimeHostPath, msbuildAssemblyPath);
         }
 
+        private static string GetMSBuildAssemblyPath(Dictionary<string, string> taskHostParameters)
+        {
+            if (taskHostParameters.TryGetValue(Constants.MSBuildAssemblyPath, out string msbuildAssemblyPath))
+            {
+                ValidateNetHostSdkVersion(msbuildAssemblyPath);
+
+                return msbuildAssemblyPath;
+            }
+
+            throw new InvalidProjectFileException("Path to MSBuild.dll is not defined. Check if used SDK supports .NET Task Host.");
```

Comment:
The most likely reason this would happen is if the user is in a VS environment (and/or MSBuild Version) that is attempting to load an assembly marked with the .NET TaskHost loading metadata, right? Can we give a more directed error in this case - something along the lines of "Task X failed to be loaded from Assembly Y because it requires the use of the MSBuild .NET Runtime Task Host, but the .NET Runtime Task Host could not be located. See <aka.ms link> for details about how to resolve this error."